tests/UnifiedSpecTests/UnifiedTestRunner.php

```diff
@@ -339,6 +289,52 @@ private function getServerVersion() : string
         throw new UnexpectedValueException('Could not determine server version');
     }
 
+    /**
+     * Return the topology type.
+     *
+     * @throws UnexpectedValueException if topology is neither single nor RS nor sharded
+     */
+    private function getTopology() : string
+    {
+        // TODO: detect load-balanced topologies once PHPLIB-671 is implemented
+        switch ($this->getPrimaryServer()->getType()) {
+            case Server::TYPE_STANDALONE:
+                return RunOnRequirement::TOPOLOGY_SINGLE;
+            case Server::TYPE_RS_PRIMARY:
+                return RunOnRequirement::TOPOLOGY_REPLICASET;
+            case Server::TYPE_MONGOS:
+                return $this->isShardedClusterUsingReplicasets()
+                    ? RunOnRequirement::TOPOLOGY_SHARDED_REPLICASET
+                    : RunOnRequirement::TOPOLOGY_SHARDED;
+            default:
+                throw new UnexpectedValueException('Toplogy is neither single nor RS nor sharded');
+        }
+    }
+
+    /**
+     * Return whether the connection is authenticated.
+     */
+    private function isAuthenticated() : bool
+    {
+        $database = $this->internalClient->selectDatabase('admin');
+        $connectionStatus = $database->command(['connectionStatus' => 1])->toArray()[0];
```

Comment:
Supported since MongoDB 2.4: https://docs.mongodb.com/v3.0/reference/command/connectionStatus/

I look forward to finding out this isn't implemented properly by the Atlas proxies :smile:

Reply:
FWIW, libmongoc checks whether the connection string contains authentication info, or if the `MONGOC_TEST_USER` env variable is set. If the proxy gives us trouble, I suggest we default to something like this as a fallback.

Reply:
I'll add a comment to this method to note that we might want to consider that.